### PR TITLE
Add sorting methods for SliceMut variant

### LINKED ISSUE #49: Sorting methods for Slice variant
Some useful methods I miss from Array Of Structs are the sorting methods, which are [`sort()`](https://doc.rust-lang.org/std/primitive.slice.html#method.sort), [`sort_by()`](https://doc.rust-lang.org/std/primitive.slice.html#method.sort_by) and [`sort_by_key()`](https://doc.rust-lang.org/std/primitive.slice.html#method.sort_by_key) for the [`slice`](https://doc.rust-lang.org/std/primitive.slice.html) primitive type. From what I've seen, the Slice variant produced by the macro does not offer any sorting methods.

I believe it is possible to implement sorting using the [`permutation`](https://crates.io/crates/permutation) crate, and adding a `#[sort]` attribute. For example, the `sort()` method would be like this (Example works if `#[sort]` is removed/commented out):
```rust
use soa_derive::StructOfArray;

#[derive(StructOfArray)]
#[soa_derive(Debug)]
struct Foo {
    // The type of the field with the `#[sort]` attribute must implement the `Ord` trait
    #[sort]
    bar: u8,
    baz: bool,
}
// --------------------------------------------------------------------------------
// The macro-generated code
impl FooSliceMut<'_> {
    fn sort(&mut self) {
        use permutation::permutation as pmut;

        let mut permutation = pmut::sort(&self.bar);

        permutation.apply_slice_in_place(&mut self.bar);
        permutation.apply_slice_in_place(&mut self.baz);
    }

    // ... Other sorting methods ...
}
// --------------------------------------------------------------------------------
// Example usage
fn main() {
    use rand::Rng;

    let mut foo_vec = FooVec::new();
    let mut rng = rand::thread_rng();

    for i in 1..=10 {
        let num = rng.gen();
        foo_vec.push(Foo { bar: num, baz: i % 2 == 0 });
    }

    println!("Before sorting: {foo_vec:#?}");

    foo_vec.as_mut_slice().sort();

    println!("After sorting: {foo_vec:#?}");
}
```

I don't have any experience with making macros, but I can try implementing it. I'm open to any other ideas or suggestions about the implementation.

The alternative to pulling in a whole crate for sorting is to zip the vectors from all fields, sort them, and then split them back, which is very costly due to multiple allocations. A middle-ground solution would be to provide sorting under a feature (eg. `sort`) to avoid increasing default dependencies.

Limitations:
- Sorting methods will sort only based on the specified field (can be solved with generating differently named sorting methods for each field or providing a `#[sort(...)]` syntax for naming)
- Sorting a Vec variant requires the `as_mut_slice()` method first, as described in the project's caveats

### PR DESCRIPTION
**Pull request by Christopher-S-25**
This PR closes #49. I only wrote a test for `sort_by()`, since `Particle` can't implement `Ord`, because `f64` only implements `PartialOrd`. I tested `sort()` and `sort_by_key()` on my project, comparing them with my handwritten sorting methods that are functional, to confirm that the implementation is correct. I'd gladly make a different test with another struct to show that `sort()` and `sort_by_key()` work properly, if necessary.

Any changes or suggestions are welcome.

### CODE REVIEW

File: soa-derive-internal/src/slice.rs [hunk 1]
```diff
@@ -632,6 +661,64 @@ pub fn derive_mut(input: &Input) -> TokenStream {
                     #slice_from_raw_parts_mut
                 }
             }
+
+            #[doc(hidden)]
+            pub fn apply_permutation(&mut self, permutation: &mut soa_derive::Permutation) {
```
Comment (Luthaf) [inline soa-derive-internal/src/slice.rs:666]: I don't think this needs to be a public function. Everything should work fine with a private function!

**Review by Luthaf — COMMENTED** [review]
Looks pretty good, thanks a lot! I only have a small comment

File: soa-derive-internal/src/slice.rs (hunk 1 above)
Comment (Christopher-S-25) [reply soa-derive-internal/src/slice.rs:666]: You are right! I didn't realize that it's essentially being called only from within the implementation of the generated code

**Review by Luthaf — APPROVED** [review]
Thanks! I'll do a new release with these changes ASAP

**Pull request merged**